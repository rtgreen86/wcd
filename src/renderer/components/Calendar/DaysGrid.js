import React, { useMemo } from 'react';

import useWeekInfo from './useWeekInfo';

export default function DaysGrid({ year, month }) {
  const { weekend, firstDay } = useWeekInfo();

  const normalizeDay = day => day === 0 ? 7 : day;

  const getISODate = dateObj => dateObj.toISOString()[0];

  const getDate = (utcYear, utcMonth, utcDate, dateObj = new Date(Date.UTC(utcYear, utcMonth, utcDate))) => ({
    date: dateObj.getUTCDate(),
    day: normalizeDay(dateObj.getUTCDay()),
    isoDate: getISODate(dateObj),
  });

  const { date: firstMDate, day: firstMDay } = getDate(year, month - 1, 1);
  const { date: lastMDate } = getDate(year, month, 0);

  const gridRows = 6;
  const gridColumns = 7;
  const gridCellsCount = gridRows * gridColumns;
  const offset = (7 + firstMDay - firstDay) % 7;

  return useMemo(() => Array.from({ length: gridCellsCount }, (item, index) => ({cellNumber: index, marks: []}))

    .map(({cellNumber, ...rest}) => ({
      cellNumber,
      date: cellNumber + 1 - offset,
      ...rest
    }))

    .map(({date, ...rest}) => ({
      date,
      visible: date >= firstMDate && date <= lastMDate,
      ...rest
    }))

    .map(({date, ...rest}) => ({ ...getDate(year, month - 1, date), ...rest }))

    .map(({marks, day, ...rest}) => ({
      marks: weekend.includes(day)
        ? ['weekend', ...marks]
        : marks,
      day,
      ...rest
    }))

    .map(({ date, ...rest }) => ({
      date,
      text: date.toString(),
      ...rest
    }))

    .map(({ visible, cellNumber, text, marks }) => (
      visible
        ? <td key={cellNumber} className={marks.join(' ')}>{text}</td>
        : <td key={cellNumber}></td>
    ))

    .reduce(({ rows, columns }, item, index, { length }) => {
      columns.push(item);
      if (index !== 0 && (index + 1) % 7 === 0 || index === length - 1) {
        rows.push(<tr key={rows.length}>{columns}</tr>);
        columns = [];
      }
      return { rows, columns };
    }, { rows: [], columns: []}).rows,

  [year, month, gridCellsCount, offset, weekend]);
}
