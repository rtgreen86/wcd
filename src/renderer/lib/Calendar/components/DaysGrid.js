import React from 'react';
import PropTypes from 'prop-types';
import useDays from '../hooks/useDays';
import { gridColumns } from '../lib/Const';

export default function DaysGrid({ year, month }) {
  const days = useDays(year, month);

  return days.reduce((grid, day, index, { length }) => {
    const { visible, cellNumber, marks, date } = day;
    const className = marks.join(' ');

    const row = grid[grid.length - 1];

    row.push(visible
      ? <td key={`cell-${cellNumber}`} className={className}>{date.toString()}</td>
      : <td key={cellNumber}></td>);

    if ((index + 1) % gridColumns === 0 || index === length - 1) {
      grid[grid.length - 1] = <tr key={`row-${grid.length}`}>{grid[grid.length - 1]}</tr>;
      grid.push([]);
    }

    return grid;
  }, [[]]);
}

DaysGrid.propTypes = {
  year: PropTypes.number,
  month: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
}
