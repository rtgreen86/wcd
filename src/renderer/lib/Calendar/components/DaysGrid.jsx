import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import useDays from '../hooks/useDays';
import { gridColumns } from '../lib/Const';

export default function DaysGrid({ year, month, marks, onClick }) {
  const days = useDays(year, month);

  const marksMap = new Map(Object.entries(marks));

  const markedDays = days.map(day => {
    if (marksMap.has(day.isoDate)) {
      day.marks = [...day.marks, ...marksMap.get(day.isoDate)]
    }
    return day;
  });

  const handleClick = useCallback((event) => {
    onClick(event.currentTarget.dataset.date);
  }, [onClick]);

  return markedDays.reduce((grid, day, index, { length }) => {
    const { visible, cellNumber, marks, date, isoDate } = day;
    const className = marks.join(' ');

    const row = grid[grid.length - 1];

    row.push(visible
      ? <td key={`cell-${cellNumber}`} className={className} data-date={isoDate} onClick={handleClick} >{date.toString()}</td>
      : <td key={cellNumber}></td>);

    if ((index + 1) % gridColumns === 0 || index === length - 1) {
      grid[grid.length - 1] = <tr key={`row-${grid.length}`}>{grid[grid.length - 1]}</tr>;
      grid.push([]);
    }

    return grid;
  }, [[]]);
}

DaysGrid.propTypes = {
  year: PropTypes.number,
  month: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
  marks: PropTypes.objectOf(PropTypes.arrayOf(PropTypes.string)),
  onClick: PropTypes.func
};

DaysGrid.defaultProps = {
  marks: {},
  onClick: () => {}
};
