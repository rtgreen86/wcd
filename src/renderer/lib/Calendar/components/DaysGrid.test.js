import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

import DaysGrid from './DaysGrid.jsx';

describe('DaysGrid', () => {
  describe.each([
    [2000, 1],
    [2000, 2],
    [2000, 4],
    [2001, 1],
    [2001, 2],
    [2001, 4],
  ])('rendered with %i-%i', (year, month) => {
    test('match snapshot', function () {
      const { asFragment } = render(<table><tbody><DaysGrid year={year} month={month} /></tbody></table>);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
