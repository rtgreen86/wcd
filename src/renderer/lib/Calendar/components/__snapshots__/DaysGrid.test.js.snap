// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`DaysGrid rendered with 2000-1 match snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td />
        <td />
        <td />
        <td />
        <td />
        <td />
        <td
          class="weekend"
          data-date="2000-01-01"
        >
          1
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2000-01-02"
        >
          2
        </td>
        <td
          class=""
          data-date="2000-01-03"
        >
          3
        </td>
        <td
          class=""
          data-date="2000-01-04"
        >
          4
        </td>
        <td
          class=""
          data-date="2000-01-05"
        >
          5
        </td>
        <td
          class=""
          data-date="2000-01-06"
        >
          6
        </td>
        <td
          class=""
          data-date="2000-01-07"
        >
          7
        </td>
        <td
          class="weekend"
          data-date="2000-01-08"
        >
          8
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2000-01-09"
        >
          9
        </td>
        <td
          class=""
          data-date="2000-01-10"
        >
          10
        </td>
        <td
          class=""
          data-date="2000-01-11"
        >
          11
        </td>
        <td
          class=""
          data-date="2000-01-12"
        >
          12
        </td>
        <td
          class=""
          data-date="2000-01-13"
        >
          13
        </td>
        <td
          class=""
          data-date="2000-01-14"
        >
          14
        </td>
        <td
          class="weekend"
          data-date="2000-01-15"
        >
          15
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2000-01-16"
        >
          16
        </td>
        <td
          class=""
          data-date="2000-01-17"
        >
          17
        </td>
        <td
          class=""
          data-date="2000-01-18"
        >
          18
        </td>
        <td
          class=""
          data-date="2000-01-19"
        >
          19
        </td>
        <td
          class=""
          data-date="2000-01-20"
        >
          20
        </td>
        <td
          class=""
          data-date="2000-01-21"
        >
          21
        </td>
        <td
          class="weekend"
          data-date="2000-01-22"
        >
          22
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2000-01-23"
        >
          23
        </td>
        <td
          class=""
          data-date="2000-01-24"
        >
          24
        </td>
        <td
          class=""
          data-date="2000-01-25"
        >
          25
        </td>
        <td
          class=""
          data-date="2000-01-26"
        >
          26
        </td>
        <td
          class=""
          data-date="2000-01-27"
        >
          27
        </td>
        <td
          class=""
          data-date="2000-01-28"
        >
          28
        </td>
        <td
          class="weekend"
          data-date="2000-01-29"
        >
          29
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2000-01-30"
        >
          30
        </td>
        <td
          class=""
          data-date="2000-01-31"
        >
          31
        </td>
        <td />
        <td />
        <td />
        <td />
        <td />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`DaysGrid rendered with 2000-2 match snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td />
        <td />
        <td
          class=""
          data-date="2000-02-01"
        >
          1
        </td>
        <td
          class=""
          data-date="2000-02-02"
        >
          2
        </td>
        <td
          class=""
          data-date="2000-02-03"
        >
          3
        </td>
        <td
          class=""
          data-date="2000-02-04"
        >
          4
        </td>
        <td
          class="weekend"
          data-date="2000-02-05"
        >
          5
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2000-02-06"
        >
          6
        </td>
        <td
          class=""
          data-date="2000-02-07"
        >
          7
        </td>
        <td
          class=""
          data-date="2000-02-08"
        >
          8
        </td>
        <td
          class=""
          data-date="2000-02-09"
        >
          9
        </td>
        <td
          class=""
          data-date="2000-02-10"
        >
          10
        </td>
        <td
          class=""
          data-date="2000-02-11"
        >
          11
        </td>
        <td
          class="weekend"
          data-date="2000-02-12"
        >
          12
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2000-02-13"
        >
          13
        </td>
        <td
          class=""
          data-date="2000-02-14"
        >
          14
        </td>
        <td
          class=""
          data-date="2000-02-15"
        >
          15
        </td>
        <td
          class=""
          data-date="2000-02-16"
        >
          16
        </td>
        <td
          class=""
          data-date="2000-02-17"
        >
          17
        </td>
        <td
          class=""
          data-date="2000-02-18"
        >
          18
        </td>
        <td
          class="weekend"
          data-date="2000-02-19"
        >
          19
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2000-02-20"
        >
          20
        </td>
        <td
          class=""
          data-date="2000-02-21"
        >
          21
        </td>
        <td
          class=""
          data-date="2000-02-22"
        >
          22
        </td>
        <td
          class=""
          data-date="2000-02-23"
        >
          23
        </td>
        <td
          class=""
          data-date="2000-02-24"
        >
          24
        </td>
        <td
          class=""
          data-date="2000-02-25"
        >
          25
        </td>
        <td
          class="weekend"
          data-date="2000-02-26"
        >
          26
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2000-02-27"
        >
          27
        </td>
        <td
          class=""
          data-date="2000-02-28"
        >
          28
        </td>
        <td
          class=""
          data-date="2000-02-29"
        >
          29
        </td>
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr>
        <td />
        <td />
        <td />
        <td />
        <td />
        <td />
        <td />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`DaysGrid rendered with 2000-4 match snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td />
        <td />
        <td />
        <td />
        <td />
        <td />
        <td
          class="weekend"
          data-date="2000-04-01"
        >
          1
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2000-04-02"
        >
          2
        </td>
        <td
          class=""
          data-date="2000-04-03"
        >
          3
        </td>
        <td
          class=""
          data-date="2000-04-04"
        >
          4
        </td>
        <td
          class=""
          data-date="2000-04-05"
        >
          5
        </td>
        <td
          class=""
          data-date="2000-04-06"
        >
          6
        </td>
        <td
          class=""
          data-date="2000-04-07"
        >
          7
        </td>
        <td
          class="weekend"
          data-date="2000-04-08"
        >
          8
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2000-04-09"
        >
          9
        </td>
        <td
          class=""
          data-date="2000-04-10"
        >
          10
        </td>
        <td
          class=""
          data-date="2000-04-11"
        >
          11
        </td>
        <td
          class=""
          data-date="2000-04-12"
        >
          12
        </td>
        <td
          class=""
          data-date="2000-04-13"
        >
          13
        </td>
        <td
          class=""
          data-date="2000-04-14"
        >
          14
        </td>
        <td
          class="weekend"
          data-date="2000-04-15"
        >
          15
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2000-04-16"
        >
          16
        </td>
        <td
          class=""
          data-date="2000-04-17"
        >
          17
        </td>
        <td
          class=""
          data-date="2000-04-18"
        >
          18
        </td>
        <td
          class=""
          data-date="2000-04-19"
        >
          19
        </td>
        <td
          class=""
          data-date="2000-04-20"
        >
          20
        </td>
        <td
          class=""
          data-date="2000-04-21"
        >
          21
        </td>
        <td
          class="weekend"
          data-date="2000-04-22"
        >
          22
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2000-04-23"
        >
          23
        </td>
        <td
          class=""
          data-date="2000-04-24"
        >
          24
        </td>
        <td
          class=""
          data-date="2000-04-25"
        >
          25
        </td>
        <td
          class=""
          data-date="2000-04-26"
        >
          26
        </td>
        <td
          class=""
          data-date="2000-04-27"
        >
          27
        </td>
        <td
          class=""
          data-date="2000-04-28"
        >
          28
        </td>
        <td
          class="weekend"
          data-date="2000-04-29"
        >
          29
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2000-04-30"
        >
          30
        </td>
        <td />
        <td />
        <td />
        <td />
        <td />
        <td />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`DaysGrid rendered with 2001-1 match snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td />
        <td
          class=""
          data-date="2001-01-01"
        >
          1
        </td>
        <td
          class=""
          data-date="2001-01-02"
        >
          2
        </td>
        <td
          class=""
          data-date="2001-01-03"
        >
          3
        </td>
        <td
          class=""
          data-date="2001-01-04"
        >
          4
        </td>
        <td
          class=""
          data-date="2001-01-05"
        >
          5
        </td>
        <td
          class="weekend"
          data-date="2001-01-06"
        >
          6
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2001-01-07"
        >
          7
        </td>
        <td
          class=""
          data-date="2001-01-08"
        >
          8
        </td>
        <td
          class=""
          data-date="2001-01-09"
        >
          9
        </td>
        <td
          class=""
          data-date="2001-01-10"
        >
          10
        </td>
        <td
          class=""
          data-date="2001-01-11"
        >
          11
        </td>
        <td
          class=""
          data-date="2001-01-12"
        >
          12
        </td>
        <td
          class="weekend"
          data-date="2001-01-13"
        >
          13
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2001-01-14"
        >
          14
        </td>
        <td
          class=""
          data-date="2001-01-15"
        >
          15
        </td>
        <td
          class=""
          data-date="2001-01-16"
        >
          16
        </td>
        <td
          class=""
          data-date="2001-01-17"
        >
          17
        </td>
        <td
          class=""
          data-date="2001-01-18"
        >
          18
        </td>
        <td
          class=""
          data-date="2001-01-19"
        >
          19
        </td>
        <td
          class="weekend"
          data-date="2001-01-20"
        >
          20
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2001-01-21"
        >
          21
        </td>
        <td
          class=""
          data-date="2001-01-22"
        >
          22
        </td>
        <td
          class=""
          data-date="2001-01-23"
        >
          23
        </td>
        <td
          class=""
          data-date="2001-01-24"
        >
          24
        </td>
        <td
          class=""
          data-date="2001-01-25"
        >
          25
        </td>
        <td
          class=""
          data-date="2001-01-26"
        >
          26
        </td>
        <td
          class="weekend"
          data-date="2001-01-27"
        >
          27
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2001-01-28"
        >
          28
        </td>
        <td
          class=""
          data-date="2001-01-29"
        >
          29
        </td>
        <td
          class=""
          data-date="2001-01-30"
        >
          30
        </td>
        <td
          class=""
          data-date="2001-01-31"
        >
          31
        </td>
        <td />
        <td />
        <td />
      </tr>
      <tr>
        <td />
        <td />
        <td />
        <td />
        <td />
        <td />
        <td />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`DaysGrid rendered with 2001-2 match snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td />
        <td />
        <td />
        <td />
        <td
          class=""
          data-date="2001-02-01"
        >
          1
        </td>
        <td
          class=""
          data-date="2001-02-02"
        >
          2
        </td>
        <td
          class="weekend"
          data-date="2001-02-03"
        >
          3
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2001-02-04"
        >
          4
        </td>
        <td
          class=""
          data-date="2001-02-05"
        >
          5
        </td>
        <td
          class=""
          data-date="2001-02-06"
        >
          6
        </td>
        <td
          class=""
          data-date="2001-02-07"
        >
          7
        </td>
        <td
          class=""
          data-date="2001-02-08"
        >
          8
        </td>
        <td
          class=""
          data-date="2001-02-09"
        >
          9
        </td>
        <td
          class="weekend"
          data-date="2001-02-10"
        >
          10
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2001-02-11"
        >
          11
        </td>
        <td
          class=""
          data-date="2001-02-12"
        >
          12
        </td>
        <td
          class=""
          data-date="2001-02-13"
        >
          13
        </td>
        <td
          class=""
          data-date="2001-02-14"
        >
          14
        </td>
        <td
          class=""
          data-date="2001-02-15"
        >
          15
        </td>
        <td
          class=""
          data-date="2001-02-16"
        >
          16
        </td>
        <td
          class="weekend"
          data-date="2001-02-17"
        >
          17
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2001-02-18"
        >
          18
        </td>
        <td
          class=""
          data-date="2001-02-19"
        >
          19
        </td>
        <td
          class=""
          data-date="2001-02-20"
        >
          20
        </td>
        <td
          class=""
          data-date="2001-02-21"
        >
          21
        </td>
        <td
          class=""
          data-date="2001-02-22"
        >
          22
        </td>
        <td
          class=""
          data-date="2001-02-23"
        >
          23
        </td>
        <td
          class="weekend"
          data-date="2001-02-24"
        >
          24
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2001-02-25"
        >
          25
        </td>
        <td
          class=""
          data-date="2001-02-26"
        >
          26
        </td>
        <td
          class=""
          data-date="2001-02-27"
        >
          27
        </td>
        <td
          class=""
          data-date="2001-02-28"
        >
          28
        </td>
        <td />
        <td />
        <td />
      </tr>
      <tr>
        <td />
        <td />
        <td />
        <td />
        <td />
        <td />
        <td />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`DaysGrid rendered with 2001-4 match snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="weekend"
          data-date="2001-04-01"
        >
          1
        </td>
        <td
          class=""
          data-date="2001-04-02"
        >
          2
        </td>
        <td
          class=""
          data-date="2001-04-03"
        >
          3
        </td>
        <td
          class=""
          data-date="2001-04-04"
        >
          4
        </td>
        <td
          class=""
          data-date="2001-04-05"
        >
          5
        </td>
        <td
          class=""
          data-date="2001-04-06"
        >
          6
        </td>
        <td
          class="weekend"
          data-date="2001-04-07"
        >
          7
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2001-04-08"
        >
          8
        </td>
        <td
          class=""
          data-date="2001-04-09"
        >
          9
        </td>
        <td
          class=""
          data-date="2001-04-10"
        >
          10
        </td>
        <td
          class=""
          data-date="2001-04-11"
        >
          11
        </td>
        <td
          class=""
          data-date="2001-04-12"
        >
          12
        </td>
        <td
          class=""
          data-date="2001-04-13"
        >
          13
        </td>
        <td
          class="weekend"
          data-date="2001-04-14"
        >
          14
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2001-04-15"
        >
          15
        </td>
        <td
          class=""
          data-date="2001-04-16"
        >
          16
        </td>
        <td
          class=""
          data-date="2001-04-17"
        >
          17
        </td>
        <td
          class=""
          data-date="2001-04-18"
        >
          18
        </td>
        <td
          class=""
          data-date="2001-04-19"
        >
          19
        </td>
        <td
          class=""
          data-date="2001-04-20"
        >
          20
        </td>
        <td
          class="weekend"
          data-date="2001-04-21"
        >
          21
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2001-04-22"
        >
          22
        </td>
        <td
          class=""
          data-date="2001-04-23"
        >
          23
        </td>
        <td
          class=""
          data-date="2001-04-24"
        >
          24
        </td>
        <td
          class=""
          data-date="2001-04-25"
        >
          25
        </td>
        <td
          class=""
          data-date="2001-04-26"
        >
          26
        </td>
        <td
          class=""
          data-date="2001-04-27"
        >
          27
        </td>
        <td
          class="weekend"
          data-date="2001-04-28"
        >
          28
        </td>
      </tr>
      <tr>
        <td
          class="weekend"
          data-date="2001-04-29"
        >
          29
        </td>
        <td
          class=""
          data-date="2001-04-30"
        >
          30
        </td>
        <td />
        <td />
        <td />
        <td />
        <td />
      </tr>
      <tr>
        <td />
        <td />
        <td />
        <td />
        <td />
        <td />
        <td />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
